File: leetcode/easy/1_two_sum/two_sum.py
from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums_dict = {}
        for i in range(len(nums)):
            if nums[i] in nums_dict:
                nums_dict[nums[i]].append(i)
            else:
                nums_dict[nums[i]] = [i]
        for i in range(len(nums)):
            needed = target - nums[i]
            if needed in nums_dict:
                if needed == nums[i]:
                    if len(nums_dict[needed]) <= 1:
                        continue
                    return [nums_dict[needed][0], nums_dict[needed][1]]
                return [i, nums_dict[needed][0]]


print(Solution().twoSum([2, 7, 11, 15], 9))
print(Solution().twoSum([3, 2, 4], 6))
print(Solution().twoSum([3, 3], 6))

File: leetcode/easy/1_two_sum/two_sum.cpp
#include <unordered_map>
#include <vector>

class Solution {
public:
  std::vector<int> twoSum(std::vector<int> &nums, int target) {
    std::unordered_map<int, std::vector<int>> map;
    for (auto i = 0; i < nums.size(); i++) {
      const auto match_loc = map.find(target - nums[i]);
      if (match_loc != map.end()) {
        return {i, match_loc->second[0]};
      }
      const auto insert_loc = map.find(nums[i]);
      if (insert_loc == map.end()) {
        const std::vector<int> v{i};
        map.emplace(nums[i], std::move(v));
      } else {
        insert_loc->second.push_back(i);
      }
    }
    return {};
  }
};
